src/Workspaces/Core/Portable/Storage/SQLite/v2/Interop/SqlConnection.cs

```diff
@@ -229,34 +261,32 @@ public int LastInsertRowId()
             => (int)NativeMethods.sqlite3_last_insert_rowid(_handle);
 
         [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", AllowCaptures = false)]
-        public Stream? ReadBlob_MustRunInTransaction(Database database, string tableName, string columnName, long rowId)
+        public Optional<Stream> ReadDataBlob_MustRunInTransaction(Database database, Table table, long rowId)
         {
-            // NOTE: we do need to do the blob reading in a transaction because of the
-            // following: https://www.sqlite.org/c3ref/blob_open.html
-            //
-            // If the row that a BLOB handle points to is modified by an UPDATE, DELETE, 
-            // or by ON CONFLICT side-effects then the BLOB handle is marked as "expired".
-            // This is true if any column of the row is changed, even a column other than
-            // the one the BLOB handle is open on. Calls to sqlite3_blob_read() and 
-            // sqlite3_blob_write() for an expired BLOB handle fail with a return code of
-            // SQLITE_ABORT.
-            if (!IsInTransaction)
-            {
-                throw new InvalidOperationException("Must read blobs within a transaction to prevent corruption!");
-            }
+            return ReadBlob_MustRunInTransaction(
+                database, table, Column.Data, rowId,
+                static (self, blobHandle) => new Optional<Stream>(self.ReadBlob(blobHandle)));
+        }
 
-            const int ReadOnlyFlags = 0;
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", AllowCaptures = false)]
+        public Optional<Checksum.HashData> ReadChecksum_MustRunInTransaction(Database database, Table table, long rowId)
+        {
+            return ReadBlob_MustRunInTransaction(
+                database, table, Column.Checksum, rowId,
+                static (self, blobHandle) =>
+                {
+                    // If the length of the blob isn't correct, then we can't read a checksum out of this.
+                    var length = NativeMethods.sqlite3_blob_bytes(blobHandle);
+                    if (length != Checksum.HashSize)
+                        return new Optional<Checksum.HashData>();
 
-            using var blob = NativeMethods.sqlite3_blob_open(_handle, database.GetName(), tableName, columnName, rowId, ReadOnlyFlags, out var result);
+                    Span<byte> bytes = stackalloc byte[Checksum.HashSize];
+                    self.ThrowIfNotOk(NativeMethods.sqlite3_blob_read(blobHandle, bytes, offset: 0));
 
-            if (result == Result.ERROR)
-            {
-                // can happen when rowId points to a row that hasn't been written to yet.
-                return null;
-            }
+                    Contract.ThrowIfFalse(MemoryMarshal.TryRead(bytes, out Checksum.HashData hashData));
```

Comment:
❗ Need to use `HashData.FromPointer` here.

https://github.com/dotnet/roslyn/blob/230d0d0caa2ec3be6039202acdfdfecaeb083fd6/src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs#L98-L102